esda/correlogram.py

```diff
@@ -0,0 +1,153 @@
+# Spatial Correlograms
+
+import geopandas as gpd
+import pandas as pd
+from joblib import Parallel, delayed
+from libpysal.cg.kdtree import KDTree
+from libpysal.weights import KNN, DistanceBand
+from libpysal.weights.util import get_points_array
+
+from .geary import Geary
+from .getisord import G
+from .moran import Moran
+
+
+def _get_autocorrelation_stat(inputs):
+    """helper function for computing parallel autocorrelation statistics
+
+    Parameters
+    ----------
+    inputs : tuple
+        tuple of (y, tree, W, statistic, STATISTIC, dist, weights_kwargs, stat_kwargs)
+
+    Returns
+    -------
+    pandas.Series
+        a pandas series with the computed autocorrelation statistic and its simulated p-value
+    """
+    (
+        y,  # y variable
+        tree,  # kd tree
+        W,  # weights class DistanceBand or KNN
+        STATISTIC,  # class of statistic (Moran, Geary, etc)
+        dist,  # threshold/k parameter for the weights
+        weights_kwargs,  # additional args
+        stat_kwargs,  # additional args
+    ) = inputs
+
+    w = W(tree, dist, silence_warnings=True, **weights_kwargs)
+    autocorr = STATISTIC(y, w, **stat_kwargs)
+    attrs = []
+    all_attrs = list(dict(vars(autocorr)).keys())
+    for attribute in all_attrs:
+        attrs.append(getattr(autocorr, str(attribute)))
+    return pd.Series(attrs, index=all_attrs, name=dist)
+
+
+def correlogram(
+    gdf: gpd.GeoDataFrame,
+    variable: str,
+    distances: list,
+    distance_type: str = "band",
+    statistic: str = "I",
+    weights_kwargs: dict = None,
+    stat_kwargs: dict = None,
+    n_jobs: int = -1,
+    backend: str = "loky",
+):
+    """Generate a spatial correlogram
+
+    A spatial correlogram is a set of spatial autocorrelation statistics calculated for
+    a set of increasing distances. It is a useful exploratory tool for examining
+    how the relationship between spatial units changes over different notions of scale.
+
+    Parameters
+    ----------
+    gdf : gpd.GeoDataFrame
+        geodataframe holding spatial and attribute data
+    variable: str
+        column on the geodataframe used to compute autocorrelation statistic
+    distances : list
+        list of distances to compute the autocorrelation statistic
+    distance_type : str, optional
+        which concept of distance to increment. Options are {`band`, `knn`}.
+        by default 'band' (for `libpysal.weights.DistanceBand` weights)
+    statistic : str, by default 'I'
+        which spatial autocorrelation statistic to compute. Options in {`I`, `G`, `C`}
```

Comment:
could you add a key to what I, G, C mean? Saying that it is Moran's I, Geary C, Getis-Ord G?

Reply:
This is a nice suggestion.